File: binding/src/spatialaudio/libqspatialsound.cpp
#include <QAudioEngine>
#include <QObject>
#include <QQuaternion>
#include <QSpatialSound>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <QVector3D>
#include <qspatialsound.h>
#include "libqspatialsound.h"
#include "libqspatialsound.hxx"

QSpatialSound* QSpatialSound_new(QAudioEngine* engine) {
	 return new QSpatialSound(engine);
}

libqt_string QSpatialSound_Tr(const char* s) {
	QString _ret = QSpatialSound::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSpatialSound_SetSource(QSpatialSound* self, const QUrl* url) {
	self->setSource(*url);
}

QUrl* QSpatialSound_Source(const QSpatialSound* self) {
	return new QUrl(self->source());
}

int QSpatialSound_Loops(const QSpatialSound* self) {
	return self->loops();
}

void QSpatialSound_SetLoops(QSpatialSound* self, int loops) {
	self->setLoops(loops);
}

bool QSpatialSound_AutoPlay(const QSpatialSound* self) {
	return self->autoPlay();
}

void QSpatialSound_SetAutoPlay(QSpatialSound* self, bool autoPlay) {
	self->setAutoPlay(autoPlay);
}

void QSpatialSound_SetPosition(QSpatialSound* self, QVector3D* pos) {
	self->setPosition(*pos);
}

QVector3D* QSpatialSound_Position(const QSpatialSound* self) {
	return new QVector3D(self->position());
}

void QSpatialSound_SetRotation(QSpatialSound* self, const QQuaternion* q) {
	self->setRotation(*q);
}

QQuaternion* QSpatialSound_Rotation(const QSpatialSound* self) {
	return new QQuaternion(self->rotation());
}

void QSpatialSound_SetVolume(QSpatialSound* self, float volume) {
	self->setVolume(volume);
}

float QSpatialSound_Volume(const QSpatialSound* self) {
	return self->volume();
}

void QSpatialSound_SetDistanceModel(QSpatialSound* self, int model) {
	self->setDistanceModel(static_cast<QSpatialSound::DistanceModel>(model));
}

int QSpatialSound_DistanceModel(const QSpatialSound* self) {
	return self->distanceModel();
}

void QSpatialSound_SetSize(QSpatialSound* self, float size) {
	self->setSize(size);
}

float QSpatialSound_Size(const QSpatialSound* self) {
	return self->size();
}

void QSpatialSound_SetDistanceCutoff(QSpatialSound* self, float cutoff) {
	self->setDistanceCutoff(cutoff);
}

float QSpatialSound_DistanceCutoff(const QSpatialSound* self) {
	return self->distanceCutoff();
}

void QSpatialSound_SetManualAttenuation(QSpatialSound* self, float attenuation) {
	self->setManualAttenuation(attenuation);
}

float QSpatialSound_ManualAttenuation(const QSpatialSound* self) {
	return self->manualAttenuation();
}

void QSpatialSound_SetOcclusionIntensity(QSpatialSound* self, float occlusion) {
	self->setOcclusionIntensity(occlusion);
}

float QSpatialSound_OcclusionIntensity(const QSpatialSound* self) {
	return self->occlusionIntensity();
}

void QSpatialSound_SetDirectivity(QSpatialSound* self, float alpha) {
	self->setDirectivity(alpha);
}

float QSpatialSound_Directivity(const QSpatialSound* self) {
	return self->directivity();
}

void QSpatialSound_SetDirectivityOrder(QSpatialSound* self, float alpha) {
	self->setDirectivityOrder(alpha);
}

float QSpatialSound_DirectivityOrder(const QSpatialSound* self) {
	return self->directivityOrder();
}

void QSpatialSound_SetNearFieldGain(QSpatialSound* self, float gain) {
	self->setNearFieldGain(gain);
}

float QSpatialSound_NearFieldGain(const QSpatialSound* self) {
	return self->nearFieldGain();
}

QAudioEngine* QSpatialSound_Engine(const QSpatialSound* self) {
	return self->engine();
}

void QSpatialSound_SourceChanged(QSpatialSound* self) {
	self->sourceChanged();
}

void QSpatialSound_Connect_SourceChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::sourceChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_LoopsChanged(QSpatialSound* self) {
	self->loopsChanged();
}

void QSpatialSound_Connect_LoopsChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::loopsChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_AutoPlayChanged(QSpatialSound* self) {
	self->autoPlayChanged();
}

void QSpatialSound_Connect_AutoPlayChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::autoPlayChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_PositionChanged(QSpatialSound* self) {
	self->positionChanged();
}

void QSpatialSound_Connect_PositionChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::positionChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_RotationChanged(QSpatialSound* self) {
	self->rotationChanged();
}

void QSpatialSound_Connect_RotationChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::rotationChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_VolumeChanged(QSpatialSound* self) {
	self->volumeChanged();
}

void QSpatialSound_Connect_VolumeChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::volumeChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_DistanceModelChanged(QSpatialSound* self) {
	self->distanceModelChanged();
}

void QSpatialSound_Connect_DistanceModelChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::distanceModelChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_SizeChanged(QSpatialSound* self) {
	self->sizeChanged();
}

void QSpatialSound_Connect_SizeChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::sizeChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_DistanceCutoffChanged(QSpatialSound* self) {
	self->distanceCutoffChanged();
}

void QSpatialSound_Connect_DistanceCutoffChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::distanceCutoffChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_ManualAttenuationChanged(QSpatialSound* self) {
	self->manualAttenuationChanged();
}

void QSpatialSound_Connect_ManualAttenuationChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::manualAttenuationChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_OcclusionIntensityChanged(QSpatialSound* self) {
	self->occlusionIntensityChanged();
}

void QSpatialSound_Connect_OcclusionIntensityChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::occlusionIntensityChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_DirectivityChanged(QSpatialSound* self) {
	self->directivityChanged();
}

void QSpatialSound_Connect_DirectivityChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::directivityChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_DirectivityOrderChanged(QSpatialSound* self) {
	self->directivityOrderChanged();
}

void QSpatialSound_Connect_DirectivityOrderChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::directivityOrderChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_NearFieldGainChanged(QSpatialSound* self) {
	self->nearFieldGainChanged();
}

void QSpatialSound_Connect_NearFieldGainChanged(QSpatialSound* self, intptr_t slot) {
    void (*slotFunc)(QSpatialSound*) = reinterpret_cast<void (*)(QSpatialSound*)>(slot);
    QSpatialSound::connect(self, &QSpatialSound::nearFieldGainChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSpatialSound_Play(QSpatialSound* self) {
	self->play();
}

void QSpatialSound_Pause(QSpatialSound* self) {
	self->pause();
}

void QSpatialSound_Stop(QSpatialSound* self) {
	self->stop();
}

libqt_string QSpatialSound_Tr2(const char* s, const char* c) {
	QString _ret = QSpatialSound::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSpatialSound_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSpatialSound::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSpatialSound_Delete(QSpatialSound* self) {
    delete self;
}

File: binding/include/spatialaudio/libqspatialsound.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qspatialsound.html
pub const qspatialsound = struct {

    /// New constructs a new QSpatialSound object.
    pub fn New(engine: ?*anyopaque) QtC.QSpatialSound {
        return qtc.QSpatialSound_new(@ptrCast(engine));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSpatialSound_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qspatialsound.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetSource(self: ?*anyopaque, url: ?*anyopaque) void {
        qtc.QSpatialSound_SetSource(@ptrCast(self), @ptrCast(url));
    }

    pub fn Source(self: ?*anyopaque, ) QtC.QUrl {
        return qtc.QSpatialSound_Source(@ptrCast(self));
    }

    pub fn Loops(self: ?*anyopaque, ) i32 {
        return qtc.QSpatialSound_Loops(@ptrCast(self));
    }

    pub fn SetLoops(self: ?*anyopaque, loops: i32) void {
        qtc.QSpatialSound_SetLoops(@ptrCast(self), loops);
    }

    pub fn AutoPlay(self: ?*anyopaque, ) bool {
        return qtc.QSpatialSound_AutoPlay(@ptrCast(self));
    }

    pub fn SetAutoPlay(self: ?*anyopaque, autoPlay: bool) void {
        qtc.QSpatialSound_SetAutoPlay(@ptrCast(self), autoPlay);
    }

    pub fn SetPosition(self: ?*anyopaque, pos: QtC.QVector3D) void {
        qtc.QSpatialSound_SetPosition(@ptrCast(self), @ptrCast(pos));
    }

    pub fn Position(self: ?*anyopaque, ) QtC.QVector3D {
        return qtc.QSpatialSound_Position(@ptrCast(self));
    }

    pub fn SetRotation(self: ?*anyopaque, q: ?*anyopaque) void {
        qtc.QSpatialSound_SetRotation(@ptrCast(self), @ptrCast(q));
    }

    pub fn Rotation(self: ?*anyopaque, ) QtC.QQuaternion {
        return qtc.QSpatialSound_Rotation(@ptrCast(self));
    }

    pub fn SetVolume(self: ?*anyopaque, volume: f32) void {
        qtc.QSpatialSound_SetVolume(@ptrCast(self), volume);
    }

    pub fn Volume(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_Volume(@ptrCast(self));
    }

    pub fn SetDistanceModel(self: ?*anyopaque, model: i32) void {
        qtc.QSpatialSound_SetDistanceModel(@ptrCast(self), @intCast(model));
    }

    pub fn DistanceModel(self: ?*anyopaque, ) i32 {
        return qtc.QSpatialSound_DistanceModel(@ptrCast(self));
    }

    pub fn SetSize(self: ?*anyopaque, size: f32) void {
        qtc.QSpatialSound_SetSize(@ptrCast(self), size);
    }

    pub fn Size(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_Size(@ptrCast(self));
    }

    pub fn SetDistanceCutoff(self: ?*anyopaque, cutoff: f32) void {
        qtc.QSpatialSound_SetDistanceCutoff(@ptrCast(self), cutoff);
    }

    pub fn DistanceCutoff(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_DistanceCutoff(@ptrCast(self));
    }

    pub fn SetManualAttenuation(self: ?*anyopaque, attenuation: f32) void {
        qtc.QSpatialSound_SetManualAttenuation(@ptrCast(self), attenuation);
    }

    pub fn ManualAttenuation(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_ManualAttenuation(@ptrCast(self));
    }

    pub fn SetOcclusionIntensity(self: ?*anyopaque, occlusion: f32) void {
        qtc.QSpatialSound_SetOcclusionIntensity(@ptrCast(self), occlusion);
    }

    pub fn OcclusionIntensity(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_OcclusionIntensity(@ptrCast(self));
    }

    pub fn SetDirectivity(self: ?*anyopaque, alpha: f32) void {
        qtc.QSpatialSound_SetDirectivity(@ptrCast(self), alpha);
    }

    pub fn Directivity(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_Directivity(@ptrCast(self));
    }

    pub fn SetDirectivityOrder(self: ?*anyopaque, alpha: f32) void {
        qtc.QSpatialSound_SetDirectivityOrder(@ptrCast(self), alpha);
    }

    pub fn DirectivityOrder(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_DirectivityOrder(@ptrCast(self));
    }

    pub fn SetNearFieldGain(self: ?*anyopaque, gain: f32) void {
        qtc.QSpatialSound_SetNearFieldGain(@ptrCast(self), gain);
    }

    pub fn NearFieldGain(self: ?*anyopaque, ) f32 {
        return qtc.QSpatialSound_NearFieldGain(@ptrCast(self));
    }

    pub fn Engine(self: ?*anyopaque, ) QtC.QAudioEngine {
        return qtc.QSpatialSound_Engine(@ptrCast(self));
    }

    pub fn SourceChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_SourceChanged(@ptrCast(self));
    }

    pub fn OnSourceChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_SourceChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LoopsChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_LoopsChanged(@ptrCast(self));
    }

    pub fn OnLoopsChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_LoopsChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn AutoPlayChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_AutoPlayChanged(@ptrCast(self));
    }

    pub fn OnAutoPlayChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_AutoPlayChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn PositionChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_PositionChanged(@ptrCast(self));
    }

    pub fn OnPositionChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_PositionChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn RotationChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_RotationChanged(@ptrCast(self));
    }

    pub fn OnRotationChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_RotationChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn VolumeChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_VolumeChanged(@ptrCast(self));
    }

    pub fn OnVolumeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_VolumeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DistanceModelChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_DistanceModelChanged(@ptrCast(self));
    }

    pub fn OnDistanceModelChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_DistanceModelChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn SizeChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_SizeChanged(@ptrCast(self));
    }

    pub fn OnSizeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_SizeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DistanceCutoffChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_DistanceCutoffChanged(@ptrCast(self));
    }

    pub fn OnDistanceCutoffChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_DistanceCutoffChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ManualAttenuationChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_ManualAttenuationChanged(@ptrCast(self));
    }

    pub fn OnManualAttenuationChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_ManualAttenuationChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn OcclusionIntensityChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_OcclusionIntensityChanged(@ptrCast(self));
    }

    pub fn OnOcclusionIntensityChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_OcclusionIntensityChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DirectivityChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_DirectivityChanged(@ptrCast(self));
    }

    pub fn OnDirectivityChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_DirectivityChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DirectivityOrderChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_DirectivityOrderChanged(@ptrCast(self));
    }

    pub fn OnDirectivityOrderChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_DirectivityOrderChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn NearFieldGainChanged(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_NearFieldGainChanged(@ptrCast(self));
    }

    pub fn OnNearFieldGainChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSpatialSound_Connect_NearFieldGainChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Play(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_Play(@ptrCast(self));
    }

    pub fn Pause(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_Pause(@ptrCast(self));
    }

    pub fn Stop(self: ?*anyopaque, ) void {
        qtc.QSpatialSound_Stop(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSpatialSound_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qspatialsound.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSpatialSound_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qspatialsound.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSpatialSound_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/spatialsound.html#types
pub const enums = struct {
    pub const Loops = enum {
        pub const Infinite: i32 = -1;
        pub const Once: i32 = 1;
    };

    pub const DistanceModel = enum {
        pub const Logarithmic: i32 = 0;
        pub const Linear: i32 = 1;
        pub const ManualAttenuation: i32 = 2;
    };

};
